client/scala/src/test/scala/ai/verta/blobs/dataset/TestVersioning.scala

```diff
@@ -0,0 +1,80 @@
+package ai.verta.blobs.dataset
+
+import ai.verta.client._
+import ai.verta.blobs._
+import ai.verta.swagger.client.HttpException
+import ai.verta.swagger._public.modeldb.versioning.model.VersioningSetTagRequestResponse
+
+import scala.concurrent.ExecutionContext
+import scala.language.reflectiveCalls
+import scala.util.{Try, Success, Failure}
+
+import org.scalatest.FunSuite
+import org.scalatest.Assertions._
+
+import com.amazonaws.regions.Regions;
+import com.amazonaws.services.s3._
+import com.amazonaws.services.s3.model._
+import collection.JavaConverters._
+
+import scala.collection.mutable.HashSet
+
+class TestVersioning extends FunSuite {
+  def fixture =
+    new {
+      val testfilePath = "s3://verta-scala-test/testdir/testfile"
+      val testfileLoc = S3Location(testfilePath).get
+
+      val testfilePath2 = "s3://verta-scala-test/testdir/testsubdir/testfile2"
+      val testfileLoc2 = S3Location(testfilePath2).get
+
+      val testdirPath = "s3://verta-scala-test/testdir/"
+      val testdirLoc = S3Location(testdirPath).get
+
+      val testsubdirPath = "s3://verta-scala-test/testdir/testsubdir/"
+      val testsubdirLoc = S3Location(testsubdirPath).get
+
+      val bucketLoc = S3Location("s3://verta-scala-test").get
+
+      val workingDir = System.getProperty("user.dir")
+      val testDir = workingDir + "/src/test/scala/ai/verta/blobs/testdir"
+      val testfile = testDir + "/testfile"
+      val testSubdir = testDir + "/testsubdir"
+      val testfile2 = testSubdir + "/testfile2"
+    }
+
+  test("The downloaded s3 file should not be corrupted") {
+    val f = fixture
+    val s3Blob = S3(f.bucketLoc).get
+    val s3PathToLocalMap = s3Blob.prepareForUpload().get.mapValues(comp => comp.filePath)
+
+    val localPath = s3PathToLocalMap.get(f.testfilePath).get
+    val pathBlob = PathBlob(localPath).get
+    assert(pathBlob.getMetadata(localPath).get.md5 equals s3Blob.getMetadata(f.testfilePath).get.md5)
```

Comment:
This is checking the two md5s are equal, right? What happens if the md5 computation wrapper that you have is wrong and returns empty strings always?

Reply:
The s3Blob's MD5 computation is actually handled by S3, so if the computation is wrong for pathBlob this would fail the test.

Reply:
ETags are not exactly the same as the MD5 in the full file

Reply:
Huh really. I was under the impression that they were the same. In fact I performed a very similar check before: 
- Upload a file to S3
- Construct PathBlob and S3 blob of this same file
- Compare the MD5 of the two blobs

Maybe the file is too small? BTW I think that Python client is also using this:
https://github.com/VertaAI/modeldb/blob/master/client/verta/verta/dataset/_s3.py#L146

Reply:
Maybe this is relevant:
https://teppen.io/2018/06/23/aws_s3_etags/

Reply:
Update: It seems there is a way to get MD5 from S3 Metadata Object:
https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/ObjectMetadata.html#getContentMD5--
The encoding is different though (not sure if it's convertable between the two?). And not sure if I should replace the MD5 field from ETag to this (considering it would break consistency with Python Client).

Reply:
It is fine to use small files and etags. I was just pointing out it's not a universal truth. But my original question remained.

Reply:
Hmm your question was if my MD5 computation is wrong (e.g always return empty string) what would happen right?
There are the following cases:
- Small file: Test would fail. ETag return the correct MD5. My computation gives the wrong MD5.
- Big file: Test would always fail (unless somehow ETag magically returns the same wrong MD5 as well).

Not sure how to verify for big file yet. This check is rather defective in that sense. But at least it won't hide the problem (i.e false positive - passing tests even though there is a bug).

Reply:
Well, there's also a third case: the code is buggy. Replace the hash computation with returning an empty string and you'll see.